File: bot/models/fees.py
"""Models for fee tracking and configuration."""

from sqlalchemy import Column, Integer, String, DateTime, Float, Boolean
from datetime import datetime
from database.db import Base


class FeeConfig(Base):
    """Global fee configuration."""
    __tablename__ = "fee_config"
    
    id = Column(Integer, primary_key=True)
    
    # Fee settings
    swap_fee_percentage = Column(Float, default=1.0)  # 1% default
    min_fee_usd = Column(Float, default=0.0)  # Minimum fee in USD
    max_fee_usd = Column(Float, default=1000.0)  # Maximum fee in USD
    
    # Fee collection wallet
    fee_collector_chain = Column(String(50), default="ethereum")
    fee_collector_address = Column(String(100), nullable=True)
    
    # Status
    is_active = Column(Boolean, default=True)
    
    # Timestamps
    created_at = Column(DateTime, default=datetime.utcnow)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)


class FeeTransaction(Base):
    """Record of fees collected."""
    __tablename__ = "fee_transactions"
    
    id = Column(Integer, primary_key=True)
    
    # User and swap reference
    user_id = Column(Integer, nullable=False)
    swap_id = Column(Integer, nullable=True)  # Reference to swap transaction
    
    # Fee details
    chain = Column(String(50), nullable=False)
    token_symbol = Column(String(20), nullable=False)
    
    # Amounts
    swap_amount = Column(Float, nullable=False)  # Original swap amount
    fee_percentage = Column(Float, nullable=False)  # Fee % at time of swap
    fee_amount = Column(Float, nullable=False)  # Fee taken
    fee_amount_usd = Column(Float, nullable=True)
    
    # Status
    collected = Column(Boolean, default=False)  # Whether fee was transferred to collector
    
    # Timestamps
    created_at = Column(DateTime, default=datetime.utcnow)


class FeeSummary(Base):
    """Daily/monthly fee summaries."""
    __tablename__ = "fee_summaries"
    
    id = Column(Integer, primary_key=True)
    
    # Period
    period_type = Column(String(10), nullable=False)  # "daily" or "monthly"
    period_date = Column(String(10), nullable=False)  # YYYY-MM-DD or YYYY-MM
    
    # Totals
    total_swaps = Column(Integer, default=0)
    total_volume_usd = Column(Float, default=0.0)
    total_fees_usd = Column(Float, default=0.0)
    
    # By chain breakdown (stored as JSON string)
    chain_breakdown = Column(String(2000), nullable=True)
    
    # Timestamps
    created_at = Column(DateTime, default=datetime.utcnow)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)

File: api-ts/src/db/schema/fees.ts
import {
	boolean,
	integer,
	numeric,
	pgTable,
	real,
	serial,
	timestamp,
	varchar,
} from 'drizzle-orm/pg-core'

export const feeConfig = pgTable('fee_config', {
	id: serial('id').primaryKey(),
	swapFeePercentage: real('swap_fee_percentage').default(1.0),
	minFeeUsd: real('min_fee_usd').default(0.0),
	maxFeeUsd: real('max_fee_usd').default(1000.0),
	feeCollectorChain: varchar('fee_collector_chain', { length: 50 }).default('ethereum'),
	feeCollectorAddress: varchar('fee_collector_address', { length: 100 }),
	isActive: boolean('is_active').default(true),
	createdAt: timestamp('created_at').defaultNow(),
	updatedAt: timestamp('updated_at').defaultNow(),
})

export const feeTransactions = pgTable('fee_transactions', {
	id: serial('id').primaryKey(),
	userId: integer('user_id').notNull(),
	swapId: integer('swap_id'),
	chain: varchar('chain', { length: 50 }).notNull(),
	tokenSymbol: varchar('token_symbol', { length: 20 }).notNull(),
	swapAmount: real('swap_amount').notNull(),
	feePercentage: real('fee_percentage').notNull(),
	feeAmount: real('fee_amount').notNull(),
	feeAmountUsd: real('fee_amount_usd'),
	collected: boolean('collected').default(false),
	createdAt: timestamp('created_at').defaultNow(),
})

export type FeeConfig = typeof feeConfig.$inferSelect
export type NewFeeConfig = typeof feeConfig.$inferInsert
export type FeeTransaction = typeof feeTransactions.$inferSelect
export type NewFeeTransaction = typeof feeTransactions.$inferInsert

export const pointsTiers = pgTable('points_tiers', {
	id: serial('id').primaryKey(),
	userId: integer('user_id').notNull().unique(),
	tier: varchar('tier', { length: 20 }).default('none'),
	qualifyingXp: integer('qualifying_xp').default(0),
	qualifyingVolumeUsd: numeric('qualifying_volume_usd', { precision: 20, scale: 2 }).default('0'),
	accumulatedRewards: numeric('accumulated_rewards', { precision: 20, scale: 8 }).default('0'),
	lastClaimAt: timestamp('last_claim_at'),
	createdAt: timestamp('created_at').defaultNow(),
	updatedAt: timestamp('updated_at').defaultNow(),
})

export const feeDiscounts = pgTable('fee_discounts', {
	id: serial('id').primaryKey(),
	userId: integer('user_id').notNull(),
	discountTier: varchar('discount_tier', { length: 20 }).default('none'),
	discountPercent: real('discount_percent').default(0.0),
	validUntil: timestamp('valid_until'),
	source: varchar('source', { length: 20 }).default('points'),
	createdAt: timestamp('created_at').defaultNow(),
	updatedAt: timestamp('updated_at').defaultNow(),
})

export type PointsTier = typeof pointsTiers.$inferSelect
export type NewPointsTier = typeof pointsTiers.$inferInsert
export type FeeDiscount = typeof feeDiscounts.$inferSelect
export type NewFeeDiscount = typeof feeDiscounts.$inferInsert

export const feeSummaries = pgTable('fee_summaries', {
	id: serial('id').primaryKey(),
	periodType: varchar('period_type', { length: 10 }).notNull(),
	periodDate: varchar('period_date', { length: 10 }).notNull(),
	totalSwaps: integer('total_swaps').default(0),
	totalVolumeUsd: real('total_volume_usd').default(0.0),
	totalFeesUsd: real('total_fees_usd').default(0.0),
	chainBreakdown: varchar('chain_breakdown', { length: 2000 }),
	createdAt: timestamp('created_at').defaultNow(),
	updatedAt: timestamp('updated_at').defaultNow(),
})

export type FeeSummary = typeof feeSummaries.$inferSelect
export type NewFeeSummary = typeof feeSummaries.$inferInsert
